### PR TITLE
Implement Jerry api get builtin object

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

File: docs/02.API-REFERENCE.md [hunk 1]
```diff
@@ -33,6 +33,26 @@ Enum that contains JerryScript API value types:
 
 *Changed in [[NEXT_RELEASE]]*: Added `JERRY_TYPE_BIGINT` value.
 
+*New in version [[NEXT_RELEASE]]*.
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md:36]: I think this line should be below

File: docs/02.API-REFERENCE.md [hunk 2]
```diff
@@ -2378,6 +2398,46 @@ main (void)
...
+[doctest]: # ()
+
+```c
+#include "jerryscript.h"
+int
+main (void)
+{
+  jerry_init (JERRY_INIT_EMPTY);
+
+  jerry_value_t array_prototype = jerry_get_builtin_object (JERRY_BUILTIN_ARRAY_PROTOTYPE_OBJECT);
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md:2429]: Free the returned value.

File: jerry-core/api/jerry.c [hunk 3]
```diff
@@ -5697,6 +5761,29 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+  // internal assert for checking that builtin id is not greater than size of the jerry_builtin_id array
+  JERRY_ASSERT (sizeof (jerry_builtin_id) / sizeof (uint8_t) >= builtin_id);
+
+  if (jerry_builtin_id[builtin_id] == ECMA_BUILTIN_ID__COUNT)
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("builtin are disabled.")));
+  }
+
+  return ecma_make_object_value (ecma_builtin_get (jerry_builtin_id[builtin_id]));
+
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5784]: Increase reference count, and remove the extra newline.

File: jerry-core/api/jerry.c [hunk 4]
```diff
@@ -5697,6 +5761,29 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
 } /* jerry_get_container_type */
 
+/**
+ * Get builtin object
+ *
+ *
+ * @return - jerry_value_t referenced to specified built-in object
+ *         - Error value if builtin is disabled
+ */
+jerry_value_t jerry_get_builtin_object (uint8_t builtin_id  /**< id of the  builtin */)
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5771]: Wrong `)`, should be before the comment.

File: jerry-core/api/jerry.c [hunk 5]
```diff
@@ -5697,6 +5761,29 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
   return JERRY_CONTAINER_TYPE_INVALID;
 } /* jerry_get_container_type */
 
+/**
+ * Get builtin object
+ *
+ *
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5767]: Delete this line

File: docs/02.API-REFERENCE.md [hunk 6]
```diff
@@ -2378,6 +2398,46 @@ main (void)
 - [jerry_container_type_t](#jerry_container_type_t)
 
 
+*New in version [[NEXT_RELEASE]]*.
+
+## jerry_get_builtin_object
+
+**Summary**
+
+Get the builtin object.
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md:2407]: Better description: `Get the specified built-in object. The function always returns with an object, unless the feature required for the built-in object is disabled.`

File: jerry-core/api/jerry.c [hunk 7]
```diff
@@ -5697,6 +5761,29 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+ * @return - jerry_value_t referenced to specified built-in object
+ *         - Error value if builtin is disabled
+ */
+jerry_value_t jerry_get_builtin_object (uint8_t builtin_id  /**< id of the  builtin */)
+{
+  jerry_assert_api_available ();
+
+  // internal assert for checking that builtin id is not greater than size of the jerry_builtin_id array
+  JERRY_ASSERT (sizeof (jerry_builtin_id) / sizeof (uint8_t) >= builtin_id);
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5776]: Check this with an `if`, and throw an error for invalid id-s.

File: jerry-core/api/jerry.c [hunk 8]
```diff
@@ -5697,6 +5761,29 @@ jerry_get_container_type (const jerry_value_t value) /**< the container object *
...
+
+  if (jerry_builtin_id[builtin_id] == ECMA_BUILTIN_ID__COUNT)
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("builtin are disabled.")));
```
Comment (zherczeg) [inline jerry-core/api/jerry.c:5780]: disabled built-in

File: tests/unit-core/test-api-get_builtin_object.c [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+#include "test-common.h"
+
+int
+main (void)
+{
+  TEST_INIT ();
+
+  jerry_init (JERRY_INIT_EMPTY);
+
+  jerry_value_t array_prototype = jerry_get_builtin_object (JERRY_BUILTIN_ARRAY_PROTOTYPE_OBJECT);
```
Comment (zherczeg) [inline tests/unit-core/test-api-get_builtin_object.c:28]: Relese the value.

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Review by rerobika — CHANGES_REQUESTED** [review]

**Comment by kisbg** [comment]
@zherczeg  Thank you for the review!

File: jerry-core/api/jerry.c [hunk 10]
```diff
@@ -124,6 +124,70 @@ static const char * const error_bigint_not_supported_p = "BigInt support is disa
 
 #endif /* ENABLED (JERRY_ERROR_MESSAGES) */
 
+/**
+ * JerryScript builtin id
+ */
+static const uint8_t jerry_builtin_id[13] =
```
[the remark this answers is not in the document]
Comment (kisbg) [reply jerry-core/api/jerry.c:130]: Thanks, keep me updated when the patch is landed, and im going to rebase this pull request based on your work!

**Comment by zherczeg** [comment]
This api would be great for #4522. We should revive it.

**Pull request closed without merging**